lib/semian/atomic_integer.rb

```diff
@@ -0,0 +1,28 @@
+module Semian
+  class AtomicInteger #:nodoc:
+    attr_accessor :value
+
+    def initialize(_name, _permissions)
+      @value = 0
+    end
+
+    def increment_by(val)
+      self.value += val
+    end
+
+    def increment
+      increment_by(1)
+    end
+
+    def execute_atomically
+      yield if block_given?
+    end
+
+    def shared?
+      false
+    end
```

Comment:
Here is the type of code I have ready for the iteration after this one for this type of purpose:
```ruby
def test_memory_is_shared
  run_test_with_atomic_integer_classes do |klass|
    next unless @integer.shared?
    integer_2 = klass.new('TestAtomicInteger', 0660)
    integer_2.value = 100
    assert_equal 100, @integer.value
    @integer.value = 200
    assert_equal 200, integer_2.value
    @integer.value = 0
    assert_equal 0, integer_2.value
  end
end

def test_memory_not_reset_when_at_least_one_worker_using_it
  run_test_with_atomic_integer_classes do |klass|
    next unless @integer.shared?

    @integer.value = 109
    integer_2 = klass.new('TestAtomicInteger', 0660)
    assert_equal @integer.value, integer_2.value
    pid = fork do
      integer_3 = klass.new('TestAtomicInteger', 0660)
      assert_equal 109, integer_3.value
      sleep
    end
    sleep 1
    Process.kill("KILL", pid)
    Process.waitall
    fork do
      integer_3 = klass.new('TestAtomicInteger', 0660)
      assert_equal 109, integer_3.value
    end
    Process.waitall
  end
end

private

def atomic_integer_classes
  @classes ||= retrieve_descendants(::Semian::AtomicInteger)
end

def run_test_with_atomic_integer_classes(klasses = atomic_integer_classes)
  klasses.each do |klass|
    begin
      @integer = klass.new('TestAtomicInteger', 0660)
      @integer.value = 0
      yield(klass)
    ensure
      @integer.destroy
    end
  end
end


```

With this you can test the structures without regard to whether it is a shared `SysVAtomicInteger`, a non-shared (disabled or OS not supporting) `SysVAtomicInteger`, or a regular `AtomicInteger`. Furthermore, you can even make a subclass like this to emulate shareness for testing correctness purposes (since on non-semaphore systems, you want to have at least one set of tests run) as long as `#shared?` is implemented and is overriden to be true:
```ruby
class FakeSysVAtomicInteger < Semian::AtomicInteger
  class << self
    attr_accessor :resources
  end
  self.resources = {}
  attr_accessor :name
  def self.new(name, _permissions)
    obj = resources[name] ||= super
    obj.name = name
    obj
  end

  def destroy
    self.class.resources.delete(@name)
    super
  end

  def shared?
    true
  end
end
```
Without something like `#shared?` it is hard to do a thorough test that captures all possible conditions without repeating code excessively.